src/sinks/elasticsearch.rs

```diff
@@ -230,6 +239,45 @@ impl HttpSink for ElasticSearchCommon {
     }
 }
 
+#[derive(Clone)]
+struct ElasticSearchRetryLogic;
+
+impl RetryLogic for ElasticSearchRetryLogic {
+    type Error = hyper::Error;
+    type Response = hyper::Response<Bytes>;
+
+    fn is_retriable_error(&self, error: &Self::Error) -> bool {
+        error.is_connect() || error.is_closed()
+    }
+
+    fn should_retry_response(&self, response: &Self::Response) -> RetryAction {
+        let status = response.status();
+
+        match status {
+            StatusCode::TOO_MANY_REQUESTS => RetryAction::Retry("Too many requests".into()),
+            StatusCode::NOT_IMPLEMENTED => {
+                RetryAction::DontRetry("endpoint not implemented".into())
+            }
+            _ if status.is_server_error() => RetryAction::Retry(
+                format!("{}: {}", status, String::from_utf8_lossy(response.body())).into(),
+            ),
+            _ if status.is_success() => {
+                let body = String::from_utf8_lossy(response.body());
```

Comment:
Just verifying, we are not reading the entire body into memory, correct?

Reply:
This does read the entire body, which is what we normally do for errors anyways so we can return them to the user. We can change this though its not super trivial.

That said, hyper always allocates a 8kb buffer per connection and we must read at least the first chunk. You can see here is where hyper reads off the io type from the tcp buffer into our memory buffer https://github.com/hyperium/hyper/blob/master/src/proto/h1/io.rs#L184

How large have you seen the bodies? From what I can tell this should only get allocated quickly to search then deallocated right away. Since there is no garbage collection this doesn't put much pressure on the agent.

Reply:
The body can be larger than the original payload becuase they send success / failure payloads for each and every record. This is a "gotcha" I've run into a number of times dealing with Elasticsearch. The error count should be at the very top of the response body, so I'd prefer we read just enouhg to determine the count and thats it.

Reply:
Peeking the body should be possible, it currently would require a bit of a complex change to our http client so we may want to do that in a different PR.

Reply:
The problem isn't reading the body, it's the "into memory" part. We just don't want to allocate a huge string.

The `elastic_responses` crate has `BulkErrorsResponse` for exactly this reason. During deserialization it discards everything but the errors so you don't get that blow up in memory use.

Reply:
We are doing this now https://github.com/timberio/vector/blob/master/src/sinks/util/http.rs#L281 which should allocate the whole body into a giant Vec<u8>, then using from_utf8_lossy shouldn't allocate (as long as its valid utf8). I am not sure how that is different? We should probably use that type anyways than just going into a string though.

Reply:
Oh, ok. If we're already reading the whole body into a vec then it makes no difference. But that's really not ideal for exactly this reason.

Reply:
Yup, we will need to change our retry logic to somehow support peeking the body, but this then requires us to be on task. Which makes this API much more complicated. We might be able to get away with it via using async/await but that would require hyper 0.13.

Reply:
Let's not get too far ahead of ourselves :smile: 

For now, this is probably fine as-is. We can test with large response bodies and see if it actually causes any problems with memory pressure.

Reply:
If we decide to forgo this please open an issue representing this so we can come back to it.